### PR TITLE
refactor: Rewrite all tests on top of testify and the standard library

### PR DESCRIPTION
**Pull request by notheotherben**
### Introduction
This PR rewrites the test framework on top of `testify/assert` and `testify/require` while removing the
dependency on `goconvey`.

### Contains
*Tell us what your pull request includes*

- [x] :radioactive: Breaking API changes
- [ ] :star: Features
  - [ ] :star2: New features
  - [x] :beetle: Fixes for existing features
  - [x] :100: Automated tests for my changes
- [ ] :books: Documentation
  - [ ] :memo: New documentation
  - [x] :bookmark_tabs: Fixes for existing documentation
- [ ] :electric_plug: Plugins
  - [ ] :star2: New plugins
  - [ ] :beetle: Fixes for existing plugins
  - [ ] :100: Automated tests for my changes

### Description
Prior to this PR, we've used `goconvey` as the test framework for this package. While `goconvey` does
work, it is rather slow, very heavy and adopts a...different approach to assertions. Moving to `testify` as our assertion library and using the standard `testing` library removes much of that weight and enables us to write tests which are much more succinct, easier to maintain, faster to run and easier to reason about.

There's also been a lot of work to simplify some of the more convoluted test suites and to factor out some more complex cases into reusable parameterized tests. Doing so should make it easier for us to add more cases to them in future and to catch potential edge cases which have been missed so far.

### CODE REVIEW

File: breadcrumbs.go [hunk 1]
```diff
@@ -147,16 +150,16 @@ func (l *breadcrumbsList) append(b Breadcrumb) {
...
-		}
+	for l.Length > l.MaxLength && l.Head != nil {
 		l.Head = l.Head.Next
 		l.Length--
 	}
 }
 
 func (l *breadcrumbsList) list() []Breadcrumb {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
```
Comment (notheotherben) [inline breadcrumbs.go]: This fixes a potential race condition when serializing the breadcrumbs.

File: breadcrumbs.go [hunk 2]
```diff
@@ -79,10 +79,13 @@ func (l *breadcrumbsList) WithSize(length int) BreadcrumbsList {
 
 	l.MaxLength = length
 
-	for l.Length > l.MaxLength {
-		if l.Head == nil {
-			break
-		}
+	if length == 0 {
```
Comment (notheotherben) [inline breadcrumbs.go]: This fixes an issue when using `.WithSize(0)` to clear the breadcrumb buffer.

File: options.go [hunk 3]
```diff
@@ -24,10 +24,10 @@ type MergeableOption interface {
 	Merge(old Option) Option
 }
 
-// A FinalizableOption exposes a Finalize() method which is called by the
+// A FinalizeableOption exposes a Finalize() method which is called by the
 // Packet builder before its value is used. This gives the option the opportunity
 // to perform any last-minute formatting and configuration.
-type FinalizableOption interface {
+type FinalizeableOption interface {
```
Comment (notheotherben) [inline options.go:30]: This is a breaking API change, however let's get the spelling fixed before more people depend on it.
